jax/_src/prng.py

```diff
@@ -0,0 +1,306 @@
+    raise TypeError(msg.format([lax.dtype(x) for x in [key1, key2, count]]))
+
+  odd_size = count.size % 2
+  if odd_size:
+    x = list(jnp.split(jnp.concatenate([count.ravel(), np.uint32([0])]), 2))
+  else:
+    x = list(jnp.split(count.ravel(), 2))
+
+  x = threefry2x32_p.bind(key1, key2, x[0], x[1])
+  out = jnp.concatenate(x)
+  assert out.dtype == np.uint32
+  return lax.reshape(out[:-1] if odd_size else out, count.shape)
+
+
+def threefry_split(key: jnp.ndarray, num: int = 2) -> jnp.ndarray:
+  return _threefry_split(key, int(num))  # type: ignore
+
+
+@partial(jit, static_argnums=(1,))
+def _threefry_split(key, num) -> jnp.ndarray:
+  counts = lax.iota(np.uint32, num * 2)
+  return lax.reshape(threefry_2x32(key, counts), (num, 2))
+
+
+def threefry_fold_in(key: jnp.ndarray, data: int) -> jnp.ndarray:
+  return _threefry_fold_in(key, jnp.uint32(data))
+
+
+@jit
+def _threefry_fold_in(key, data):
+  return threefry_2x32(key, threefry_init(data))
+
+
+@partial(jit, static_argnums=(1, 2))
+def threefry_random_bits(key, bit_width, shape):
+  """Sample uniform random bits of given width and shape using PRNG key."""
+  if not _is_prng_key(key):
+    raise TypeError("random_bits got invalid prng key.")
+  if bit_width not in (8, 16, 32, 64):
+    raise TypeError("requires 8-, 16-, 32- or 64-bit field width.")
+  shape = core.as_named_shape(shape)
+  for name, size in shape.named_items:
+    real_size = lax.psum(1, name)
+    if real_size != size:
+      raise ValueError(f"The shape of axis {name} was specified as {size}, "
+                       f"but it really is {real_size}")
+    axis_index = lax.axis_index(name)
+    key = threefry_fold_in(key, axis_index)
+  size = prod(shape.positional)
+  max_count = int(np.ceil(bit_width * size / 32))
+
+  nblocks, rem = divmod(max_count, jnp.iinfo(np.uint32).max)
+  if not nblocks:
+    bits = threefry_2x32(key, lax.iota(np.uint32, rem))
+  else:
+    *subkeys, last_key = threefry_split(key, nblocks + 1)
+    blocks = [threefry_2x32(k, lax.iota(np.uint32, jnp.iinfo(np.uint32).max))
+              for k in subkeys]
+    last = threefry_2x32(last_key, lax.iota(np.uint32, rem))
+    bits = lax.concatenate(blocks + [last], 0)
+
+  dtype = UINT_DTYPES[bit_width]
+  if bit_width == 64:
+    bits = [lax.convert_element_type(x, dtype) for x in jnp.split(bits, 2)]
+    bits = lax.shift_left(bits[0], dtype(32)) | bits[1]
+  elif bit_width in [8, 16]:
+    # this is essentially bits.view(dtype)[:size]
+    bits = lax.bitwise_and(
+      np.uint32(np.iinfo(dtype).max),
+      lax.shift_right_logical(
+        lax.broadcast(bits, (1,)),
+        lax.mul(
+          np.uint32(bit_width),
+          lax.broadcasted_iota(np.uint32, (32 // bit_width, 1), 0)
+        )
+      )
+    )
+    bits = lax.reshape(bits, (np.uint32(max_count * 32 // bit_width),), (1, 0))
+    bits = lax.convert_element_type(bits, dtype)[:size]
+  return lax.reshape(bits, shape)
+
+
+class threefry_prng:
```

Comment:
Should there be an abstract class or some way to enforce that implementers must implement init, fold_in, random_bits, and split? Additionally document what the contract is for each of the functions.

Reply:
Yes, we're approaching something like that. See the updated draft.